src/WebJobs.Script/Description/FunctionInvokerBase.cs

```diff
@@ -79,7 +83,79 @@ protected bool InitializeFileWatcherIfEnabled()
             return false;
         }
 
-        public abstract Task Invoke(object[] parameters);
+        public Task Invoke(object[] parameters)
+        {
+            FunctionStartedEvent startedEvent = null;
+            Task invokeTask = null;
+
+            CancellationToken timeoutOrShutdownToken = (CancellationToken)parameters.First(p => p?.GetType() == typeof(CancellationToken));
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters.First(p => p?.GetType() == typeof(ExecutionContext));
+            string invocationId = functionExecutionContext.InvocationId.ToString();
+
+            startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                TraceWriter.Info($"Function started (Id={invocationId})");
+
+                invokeTask = InvokeInternal(parameters);
+                invokeTask.Wait(timeoutOrShutdownToken);
+
+                TraceWriter.Info($"Function completed (Success, Id={invocationId})");
+            }
+            catch (Exception ex)
+            {
+                OperationCanceledException cancelledEx = ex as OperationCanceledException;
+                if (cancelledEx?.CancellationToken == timeoutOrShutdownToken)
+                {
+                    // TODO: What's the best way to do this?
+                    var stateProp = typeof(JobHost).GetField("_state", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+                    int hostState = (int)stateProp.GetValue(Host);
```

Comment:
I still don't have a good way to tell when there is a timeout. Right now if the token fires, it can be either due to a host shutdown or a timeout. I am checking private state to see if it's a shutdown.

I still think we need a proper 'Timeout' event coming from the host so we can be sure.